src/dmd/backend/symbol.d

```diff
@@ -0,0 +1,2454 @@
+    printf(" Sxtrnnum = %d\n",s.Sxtrnnum);
+    printf("  Stype   = %p",s.Stype);
+version (SCPP_HTOD)
+{
+    printf(" Ssequence = %x", s.Ssequence);
+    printf(" Scover  = %p", s.Scover);
+}
+    printf(" Sl      = %p",s.Sl);
+    printf(" Sr      = %p\n",s.Sr);
+    if (s.Sscope)
+        printf(" Sscope = '%s'\n",s.Sscope.Sident.ptr);
+    if (s.Stype)
+        type_print(s.Stype);
+    if (s.Sclass == SCmember || s.Sclass == SCfield)
+    {
+        printf("  Smemoff =%5lld", cast(long)s.Smemoff);
+        printf("  Sbit    =%3d",s.Sbit);
+        printf("  Swidth  =%3d\n",s.Swidth);
+    }
+version (SCPP_HTOD)
+{
+    if (s.Sclass == SCstruct)
+    {
+        printf("  Svbptr = %p, Svptr = %p\n",s.Sstruct.Svbptr,s.Sstruct.Svptr);
+    }
+}
+}
+}
+}
+
+
+/*********************************
+ * Terminate use of symbol table.
+ */
+
+private __gshared Symbol *keep;
+
+void symbol_term()
+{
+    symbol_free(keep);
+}
+
+/****************************************
+ * Keep symbol around until symbol_term().
+ */
+
+static if (TERMCODE)
+{
+
+void symbol_keep(Symbol *s)
+{
+    symbol_debug(s);
+    s.Sr = keep;       // use Sr so symbol_free() doesn't nest
+    keep = s;
+}
+
+}
+
+/****************************************
+ * Return alignment of symbol.
+ */
+int Symbol_Salignsize(Symbol* s)
+{
+    if (s.Salignment > 0)
+        return s.Salignment;
+    int alignsize = type_alignsize(s.Stype);
+
+    /* Reduce alignment faults when SIMD vectors
+     * are reinterpreted cast to other types with less alignment.
+     */
+    if (config.fpxmmregs && alignsize < 16 &&
+        s.Sclass == SCauto &&
+        type_size(s.Stype) == 16)
+    {
+        alignsize = 16;
+    }
+
+    return alignsize;
+}
+
+/****************************************
+ * Aver if Symbol is not only merely dead, but really most sincerely dead.
+ * Params:
+ *      anyInlineAsm = true if there's any inline assembler code
+ * Returns:
+ *      true if symbol is dead.
+ */
+
+bool Symbol_Sisdead(Symbol* s, bool anyInlineAsm)
+{
+    version (MARS)
+        enum vol = false;
+    else
+        enum vol = true;
+    return s.Sflags & SFLdead ||
+           /* SFLdead means the optimizer found no references to it.
+            * The rest deals with variables that the compiler never needed
+            * to read from memory because they were cached in registers,
+            * and so no memory needs to be allocated for them.
+            * Code that does write those variables to memory gets NOPed out
+            * during address assignment.
+            */
+           (!anyInlineAsm && !(s.Sflags & SFLread) && s.Sflags & SFLunambig &&
+
+            // mTYvolatile means this variable has been reference by a nested function
+            (vol || !(s.Stype.Tty & mTYvolatile)) &&
+
+            (config.flags4 & CFG4optimized || !config.fulltypes));
+}
+
+/****************************************
+ * Determine if symbol needs a 'this' pointer.
+ */
+
+int Symbol_needThis(Symbol* s)
```

Comment:
This looks new.

Reply:
That's because the original was Symbol::needThis(). D doesn't allow member functions to be defined outside of the aggregate, so this was the easiest translation.